compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -3182,6 +3182,9 @@ public void visit(BLangSimpleVarRef varRefExpr, AnalyzerData data) {
             // TODO: call to isInLocallyDefinedRecord() is a temporary fix done to disallow local var references in
             //  locally defined record type defs. This check should be removed once local var referencing is supported.
             if (((symbol.tag & SymTag.VARIABLE) == SymTag.VARIABLE)) {
+                if (data.commonAnalyzerData.withinAnnotationExpr && varName.getValue().equals("self")) {
```

Comment:
IMO, this is not correct. Before reaching this logic, we check for `self` within the main scope and set the symbol for self, which means we are allowing the self variable within that scope. This is not correct. At line 3171(`lookupMainSpaceSymbolInPackage`), when checking for the symbol within the scope, we should not see self for this annotation scope. 
What we did here is set the symbol of the variable and then handle it as a special case.